lib/cn/cn_tree.c

```diff
@@ -2427,64 +2447,51 @@ cn_comp_spill(struct cn_compaction_work *w)
 /**
  * cn_comp_compact() - perform the actual compaction operation
  * See section comment for more info.
+ *
+ * [HSE_REVISIT] If a spill/zspill action leaves this function too early
+ * it will prevent proper advance of the spill gen and potentially leave
+ * ongoing and/or future spill threads stuck in cn_node_spill_wait().
  */
 static void
 cn_comp_compact(struct cn_compaction_work *w)
 {
-    bool kcompact = (w->cw_action == CN_ACTION_COMPACT_K);
     struct kvdb_health *hp = w->cw_tree->ct_kvdb_health;
-    bool can_zspill = false;
-
     merr_t err = 0;
 
-    if (w->cw_err)
-        return;
-
     assert(hp);
+    assert(!w->cw_err);
 
-    w->cw_err = kvdb_health_check(hp, KVDB_HEALTH_FLAG_ALL);
-    if (w->cw_err)
-        return;
-
-    w->cw_keep_vblks = kcompact;
-    hp = w->cw_tree->ct_kvdb_health;
-    assert(hp);
-
-    if (w->cw_action == CN_ACTION_ZSPILL) {
-
-        w->cw_horizon = cn_get_seqno_horizon(w->cw_tree->cn);
-        w->cw_cancel_request = cn_get_cancel(w->cw_tree->cn);
+    w->cw_keep_vblks = (w->cw_action == CN_ACTION_COMPACT_K);
 
-        perfc_inc(w->cw_pc, PERFC_BA_CNCOMP_START);
+    w->cw_horizon = cn_get_seqno_horizon(w->cw_tree->cn);
+    w->cw_cancel_request = cn_get_cancel(w->cw_tree->cn);
 
-        cn_setname(w->cw_threadname);
+    perfc_inc(w->cw_pc, PERFC_BA_CNCOMP_START);
 
-        w->cw_err = cn_comp_zspill(w, &can_zspill);
-        if (w->cw_err)
-            return;
+    if (w->cw_action == CN_ACTION_ZSPILL) {
+        bool can_zspill;
 
+        /* If the conditions for zspill are no longer true then we must
+         * downgrade the operation to an rspill in order to properly
+         * advance the spill gen.
+         */
+        can_zspill = cn_comp_zspill(w);
         if (!can_zspill) {
+            log_info("no can do %u, %lu", w->cw_kvset_cnt, w->cw_dgen_hi_min);
             w->cw_action = CN_ACTION_SPILL;
             w->cw_rule = CN_RULE_RSPILL;
         }
     }
 
-    if (w->cw_action != CN_ACTION_ZSPILL) {
-        w->cw_err = cn_tree_prepare_compaction(w);
-        if (w->cw_err) {
-            if (merr_errno(w->cw_err) != ESHUTDOWN)
-                kvdb_health_error(hp, w->cw_err);
-            return;
-        }
+    w->cw_err = cn_tree_prepare_compaction(w);
+    if (w->cw_err) {
+        if (merr_errno(w->cw_err) != ESHUTDOWN)
+            kvdb_health_error(hp, w->cw_err);
+        return;
     }
 
     w->cw_t2_prep = get_time_ns();
 
-    /* cn_kcompact handles k-compaction, cn_spill handles spills
-     * and kv-compaction.
-     */
-    w->cw_keep_vblks = (w->cw_action == CN_ACTION_COMPACT_K);
-
```

Comment:
Where did this go?  Into code specific to kcompact mabe?

Reply:
There were two lines setting cw_keep_vblocks in this function.  This is the second, seemed unnecessary, unless prepare_compacting() is clearing it???  Seems like we could just move it into the switch case for kcompact, or set it in sp3_work() ???

Reply:
Look on line 2464 of the changed file

Reply:
I was being lazy.

Reply:
A while back I think I was looking at removing this struct member and just using `action == COMPACT_K` where needed.  Worth considering....